shell/platform/linux/fl_renderer.cc

```diff
@@ -474,10 +483,73 @@ gboolean fl_renderer_present_layers(FlRenderer* self,
 
   FlView* view =
       FL_VIEW(g_hash_table_lookup(priv->views, GINT_TO_POINTER(view_id)));
-  if (view != nullptr) {
-    fl_view_redraw(view);
+  if (view == nullptr) {
+    return TRUE;
+  }
+
+  if (view_id == flutter::kFlutterImplicitViewId) {
+    // Store for rendering later
+    g_hash_table_insert(priv->framebuffers_by_view_id, GINT_TO_POINTER(view_id),
+                        g_ptr_array_ref(framebuffers));
+  } else {
+    // Composite into a single framebuffer.
+    if (framebuffers->len > 1) {
+      size_t width = 0, height = 0;
+
+      for (guint i = 0; i < framebuffers->len; i++) {
+        FlFramebuffer* framebuffer =
+            FL_FRAMEBUFFER(g_ptr_array_index(framebuffers, i));
+
+        size_t w = fl_framebuffer_get_width(framebuffer);
+        size_t h = fl_framebuffer_get_height(framebuffer);
+        if (w > width) {
+          width = w;
+        }
+        if (h > height) {
+          height = h;
+        }
+      }
+
+      FlFramebuffer* view_framebuffer =
+          fl_framebuffer_new(priv->general_format, width, height);
+      glBindFramebuffer(GL_DRAW_FRAMEBUFFER,
```

Comment:
What's Flutter Linux's best practices around OpenGL error reporting? Should we be checking for errors after GL calls?

On Flutter Windows we started migrating to Impeller's OpenGL abstractions. Specifically, we use their [GLES proc table](https://github.com/flutter/engine/blob/main/impeller/renderer/backend/gles/proc_table_gles.h) to get conveniences like automatic error reporting. Here's a Flutter Windows [example](https://github.com/flutter/engine/blob/9513c54456927336666df817a0fa13882ea5faea/shell/platform/windows/compositor_opengl.cc#L39-L52). This has been really useful for debugging weird GL errors. In the future, should Flutter Linux move to this as well?

Reply:
You can turn on debugging using the `GDK_DEBUG=opengl` environment variable (this is built into GTK). I think this is sufficient if you are aware of any issue and are investigating.